src/streamer.rs

```diff
@@ -0,0 +1,338 @@
+    pub size: usize,
+    pub addr: [u16; 8],
+    pub port: u16,
+    pub v6: bool,
+}
+
+impl Default for Packet {
+    fn default() -> Packet {
+        Packet {
+            data: [0u8; PACKET_SIZE],
+            size: 0,
+            addr: [0u16; 8],
+            port: 0,
+            v6: false,
+        }
+    }
+}
+impl Packet {
+    pub fn get_addr(&self) -> SocketAddr {
+        if !self.v6 {
+            let ipv4 = Ipv4Addr::new(
+                self.addr[0] as u8,
+                self.addr[1] as u8,
+                self.addr[2] as u8,
+                self.addr[3] as u8,
+            );
+            SocketAddr::new(IpAddr::V4(ipv4), self.port)
+        } else {
+            let ipv6 = Ipv6Addr::new(
+                self.addr[0],
+                self.addr[1],
+                self.addr[2],
+                self.addr[3],
+                self.addr[4],
+                self.addr[5],
+                self.addr[6],
+                self.addr[7],
+            );
+            SocketAddr::new(IpAddr::V6(ipv6), self.port)
+        }
+    }
+
+    pub fn set_addr(&mut self, a: &SocketAddr) {
+        match a {
+            &SocketAddr::V4(v4) => {
+                let ip = v4.ip().octets();
+                self.addr[0] = ip[0] as u16;
+                self.addr[1] = ip[1] as u16;
+                self.addr[2] = ip[2] as u16;
+                self.addr[3] = ip[3] as u16;
+                self.port = a.port();
+            }
+            &SocketAddr::V6(v6) => {
+                self.addr = v6.ip().segments();
+                self.port = a.port();
+                self.v6 = true;
+            }
+        }
+    }
+}
+
+pub struct PacketData {
+    pub packets: Vec<Packet>,
+}
+
+type SharedPacketData = Arc<RwLock<PacketData>>;
+type Recycler = Arc<Mutex<Vec<SharedPacketData>>>;
+
+impl PacketData {
+    pub fn new() -> PacketData {
+        PacketData {
+            packets: vec![Packet::default(); BLOCK_SIZE],
+        }
+    }
+    fn run_read_from(&mut self, socket: &UdpSocket) -> Result<usize> {
+        self.packets.resize(BLOCK_SIZE, Packet::default());
+        let mut i = 0;
+        for p in self.packets.iter_mut() {
+            p.size = 0;
+            match socket.recv_from(&mut p.data) {
+                Err(_) if i > 0 => {
+                    break;
+                }
+                Err(e) => {
+                    info!("recv_from err {:?}", e);
+                    return Err(Error::IO(e));
+                }
+                Ok((nrecv, from)) => {
+                    p.size = nrecv;
+                    p.set_addr(&from);
+                    if i == 0 {
+                        socket.set_nonblocking(true)?;
+                    }
+                }
+            }
+            i += 1;
+        }
+        return Ok(i);
+    }
+    fn read_from(&mut self, socket: &UdpSocket) -> Result<()> {
+        let sz = self.run_read_from(socket)?;
+        self.packets.resize(sz, Packet::default());
+        return Ok(());
```

Comment:
The last line of a function should just be an expression, not a statement. In most of you functions, can you replace `return Ok(());` with just `Ok(())`?

Reply:
yar, you cant teach an old c-dog new tricks